dbuf-lsp/src/main.rs

```diff
@@ -1,38 +1,66 @@
 use tower_lsp::jsonrpc::Result;
+use tower_lsp::lsp_types::request::*;
 use tower_lsp::lsp_types::*;
 use tower_lsp::{Client, LanguageServer, LspService, Server};
 
-#[allow(unused_imports)]
-#[allow(clippy::single_component_path_imports)]
-use dbuf_core;
+use dbuf_lsp::handler::Handler;
+use dbuf_lsp::WorkspaceAccess;
+
+use dbuf_lsp::action_handler::ActionHandler;
+use dbuf_lsp::completion_handler::CompletitionHandler;
+use dbuf_lsp::diagnostic_handler::DiagnosticHandler;
+use dbuf_lsp::navigation_handler::NavigationHandler;
 
-#[derive(Debug)]
-#[allow(dead_code)]
 struct Backend {
     client: Client,
+    workspace: WorkspaceAccess,
+    action_handler: ActionHandler,
+    completition_handler: CompletitionHandler,
+    diagnostic_handler: DiagnosticHandler,
+    navigation_handler: NavigationHandler,
+}
 
-    ast: Option<()>,
+impl Backend {
+    fn new(client: Client) -> Self {
+        Self {
+            client,
+            workspace: WorkspaceAccess::new(),
+            action_handler: ActionHandler::new(),
+            completition_handler: CompletitionHandler::new(),
+            diagnostic_handler: DiagnosticHandler::new(),
+            navigation_handler: NavigationHandler::new(),
+        }
+    }
 }
 
 #[tower_lsp::async_trait]
 impl LanguageServer for Backend {
-    async fn initialize(&self, _: InitializeParams) -> Result<InitializeResult> {
-        eprintln!("init");
+    async fn initialize(&self, init: InitializeParams) -> Result<InitializeResult> {
+        let mut capabilities = ServerCapabilities {
+            text_document_sync: Some(TextDocumentSyncCapability::Options(
+                TextDocumentSyncOptions {
+                    open_close: Some(true),
+                    change: Some(TextDocumentSyncKind::FULL),
+                    will_save: Some(false),
+                    will_save_wait_until: Some(false),
+                    save: Some(TextDocumentSyncSaveOptions::Supported(false)),
+                },
+            )),
+            ..Default::default()
+        };
+
+        self.action_handler.init(&init, &mut capabilities);
```

Comment:
I don't like that each handler has complete access to the whole `capabilities` struct, maybe each of them returns some kind of their own dedicated "handler capability" from which you construct the whole server capability?